### ISSUE TITLE
RFE: prevent services from spawning further namespaces

### ISSUE THREAD

**Issue by lucab** [systemd/systemd#4043]
### Submission type
- Request for enhancement (RFE)
### systemd version the issue has been seen with

v231 - HEAD
### Description

With the introduction of user namespaces, unprivileged processes can create their own namespace where they have (segregated-)root privileges and further interact with the kernel. There have been concerns from multiple audiences about the theoretical attack surface this opens, and several non-mainstreamed attempt at forbidding this:
- grsec [disables them](https://forums.grsecurity.net/viewtopic.php?f=3&t=3929)
- debian has a global sysctl knob [to disable them](https://anonscm.debian.org/cgit/kernel/linux.git/tree/debian/patches/debian/add-sysctl-to-disallow-unprivileged-CLONE_NEWUSER-by-default.patch)
- there was an [attempt](https://lwn.net/Articles/673597/) at upstreaming a similar `userns_restrict` knob

Most of those focus on disabling them globally, but an alternative approach could be to selectively prevent some processes from using them, as many traditional services don't directly manipulate namespaces (and systemd can anyway take care of the segregation). As such I'm wondering if it makes sense to have a unit property or something to forbid this (eg. via seccomp).

Usecases:
- further limiting the attack surface in case of exploits against unprivileged services
- preventing already containerized/namespaced application from fiddling/nesting sub-namespaces

Notes:
1. while it should be possible to do it right now via raw/custom seccomp filters, `SystemCallFilter=` syntax can't cope with complex args filtering to surgically block `clone()`.
2. proper support may (or may not) appear at some later point in kernel (eg. via a resource limit or a namespaced sysctl knob) and it could be desirable to then re-wire this feature.

**Comment by lucab** [comment]
I realized only after submitting this that there is already a [TODO entry](https://github.com/systemd/systemd/blob/master/TODO#L73) for `RestrictNamespaces=`. So I guess this is quite a valuable feature and I can ask for more details regarding:
- scope (all namespaces? userns only?)
- property naming and syntax
- implementation sketches (seccomp-based is the first one that come to my mind, but perhaps there exist alternatives here)

**Comment by poettering** [comment]
I'd love to see that implemented!

One possible design could be a new option RestrictNamespaces= that takes either a bool, or a series of namespace identifiers. Examples:

```
RestrictNamespaces=no              # Not restrictions, the default
RestrictNamespaces=yes             # Full restrictions, no namespaces are allowed
RestrictNamespaces=pid uts         # Only permit PID and UTS namespaces
RestrictNamespaces=~user pid       # Permit all kinds of namespaces, except for pid and user
```

Internally this should just map to a clone flags field, that we enforce.

The "~" logic would match what SystemCallFilters= and the other options implement.

And seccomp-based filtering should be sufficient to implement this, as we could match against the parameters for clone(), setns() and unshare() that way. However there's one limitation: setns() permits passing 0 as flags field, in which case the namespace type is determined automatically. I figure we should prohibit this mode of invocation if this option is used. Programs that use setns() with a 0 argument, should probably not use this option at all, or should be fixed to specify the right mask. Hence I think this limitation is acceptable.

**Comment by poettering** [comment]
Hmm, one more thing: I think that CLONE_NEWNS is stupidly named, which is why I think we should expose this under a different name, that clarifies that this is about file system namespaces. However we probably should accept "ns" as identifier for this too (as an alias), so that using the kernel's own identifiers are supported for compat. Or to summarize this, I think RestrictNamespaces= should accept the following identifiers:

```
CLONE_NEWCGROUP → "cgroup"
CLONE_NEWIPC → "ipc"
CLONE_NEWNET → "net"
CLONE_NEWNS → "filesystem" (with an alias of "ns")
CLONE_NEWPID → "pid"
CLONE_NEWUSER → "user"
CLONE_NEWUTS → "uts"
```

**Comment by lucab** [comment]
@kees pointed me to the patchset at https://git.kernel.org/cgit/linux/kernel/git/ebiederm/user-namespace.git/log/?h=for-testing, which should hopefully land upstream in few release cycles.

It introduces user-namespaced sysctl knobs to limit the maximum number of ns to be created, with 0 disabling creation.

Should it be taken into account here and how?

**Comment by poettering** [comment]
This has been implemented in #4536 btw. Closing.
